Пишем первую функцию

Petrusia opened the pull request:
---
:mortar_board: [Пишем первую функцию](https://up.htmlacademy.ru/php/12/user/147450/tasks/6)

Petrusia committed:
Заменил на  одинарные кавычки

Petrusia committed:
Пишем первую функцию

Petrusia committed:
Поправил первую функцию

Petrusia committed:
Поправил первую функцию 2

Petrusia committed:
поменял strlen  на mb_strlen

olegmifle commented on helpers/truncate.php:
```diff
@@ -0,0 +1,47 @@
+/* функция, которая обрезает текстовое содержимое
+если оно превышает заданное число символов */
+
+function truncate($description, $length = 300)
```
Нужно добавить типизацию
```suggestion
function truncate(string $description, int $length = 300): string
```

olegmifle commented on helpers/truncate.php:
```diff
@@ -0,0 +1,47 @@
+    $description = trim($description);
+
+    if (mb_strlen($description) <= $length) {
+        return "<p>$description</p>";
```
Обычно для конкатенации я рекомендую использовать sprintf и vsprintf т.к. их лучше видно в коде. Если используешь двойные кавычки, то лучше выделять переменны фигурными скобками. `"<p>{$description}</p>"`
Но так лучше на мой взгляд
`sprintf('<p>%s</p>', $description)`

olegmifle commented on helpers/truncate.php:
```diff
@@ -0,0 +1,47 @@
+    $count = 0;
+
+    // разбил текст на отдельные слова (по пробелам)
+    $description = explode(' ', $description);
```
Будут ли в тексте html теги? С обрезанием текста с html тегами нужно быть очень аккуратным. Есть риск отрезать закрывающий тег.

olegmifle commented on helpers/truncate.php:
```diff
@@ -0,0 +1,47 @@
+    $description = explode(' ', $description);
+
+    //  в цикле последовательно считаю длину каждого слова и пробела
+    foreach ($description as $key => $word) {
+        $count += mb_strlen($word) + 1; // + 1 это пробел
+
+        /* останавливаю  цикл, когда суммарная длина символов
+         в посчитанных словах начинает превышать ограничение */
+        if ($count > $length) {
+            // обрезаю до приемлемой длины
+            $abbreviation = array_slice($description, 0, $key);
+            //  сложил  отдельные слова обратно в строку
+            $description = implode(' ', $abbreviation);
+            return "<p> $description $append </p> $read_more";
+        }
+    }
+}
```
Можно немного улучшить читаемость. Например, отделить неподходящие кейсы. Например
```
if ($count <= $length) { 
    continue;
}

//some action
```

Petrusia committed:
Как предложил Олег

Petrusia committed:
Как предложил Олег,
и к ключу я добавил единицу, потому что массив отсчитывает от 0, а слова - от 1, вывод 283 символа, еще одно слово может поместиться.

olegmifle commented on helpers/truncate.php:
```diff
@@ -0,0 +1,46 @@
+    $description = trim($description);
+
+    if (mb_strlen($description) <= $length) {
+        // return "<p>{$description}</p>";
```
Это лишнее
